lib/api/apiUtils/authorization/permissionChecks.js

```diff
@@ -0,0 +1,302 @@
+            return true;
+        }
+    }
+
+    // User is already authorized on the bucket for FULL_CONTROL or WRITE or
+    // bucket has canned ACL public-read-write
+    if (requestType === 'objectPut' || requestType === 'objectDelete') {
+        return true;
+    }
+
+    if (requestType === 'objectPutACL') {
+        if ((objectMD.acl.Canned === 'bucket-owner-full-control'
+                && bucketOwner === canonicalID)
+            || objectMD.acl.FULL_CONTROL.indexOf(canonicalID) > -1
+            || objectMD.acl.WRITE_ACP.indexOf(canonicalID) > -1) {
+            return true;
+        }
+    }
+
+    if (requestType === 'objectGetACL') {
+        if ((objectMD.acl.Canned === 'bucket-owner-full-control'
+                && bucketOwner === canonicalID)
+            || objectMD.acl.FULL_CONTROL.indexOf(canonicalID) > -1
+            || objectMD.acl.READ_ACP.indexOf(canonicalID) > -1) {
+            return true;
+        }
+    }
+
+    // allow public reads on buckets that are whitelisted for anonymous reads
+    // TODO: remove this after bucket policies are implemented
+    const bucketAcl = bucket.getAcl();
+    const allowPublicReads = publicReadBuckets.includes(bucket.getName()) &&
+        bucketAcl.Canned === 'public-read' &&
+        (requestType === 'objectGet' || requestType === 'objectHead');
+    if (allowPublicReads) {
+        return true;
+    }
+    return false;
+}
+
+function _checkAction(requestType, action) {
+    // if requestType isn't in list of controlled actions
+    if (!Object.values(actionMap).includes(requestType)) {
+        return true;
+    }
+    if (action === 's3:*') {
+        return true;
+    }
+    if (requestType === actionMap[action]) {
+        return true;
+    }
+    if (Array.isArray(action)
+      && action.some(a => requestType === actionMap[a])) {
+        return true;
+    }
+    return false;
+}
+
+function _checkPrincipal(canonicalID, arn, principal) {
+    if (principal === '*') {
+        return true;
+    }
+    if (principal.CanonicalUser && principal.CanonicalUser === canonicalID) {
+        return true;
+    }
+    if (Array.isArray(principal.CanonicalUser)
+    && principal.CanonicalUser.includes(canonicalID)) {
+        return true;
+    }
+    if (principal.AWS && (principal.AWS === '*' || principal.AWS === arn)) {
+        return true;
+    }
+    if (Array.isArray(principal.AWS) && principal.AWS.includes(arn)) {
+        return true;
+    }
+    return false;
+}
+
+function checkBucketPolicy(policy, requestType, canonicalID, arn) {
+    let allowed = false;
+    let explicitDeny = false;
+    while (!explicitDeny && policy.Statement.length > 0) {
+        const s = policy.Statement[0];
+        const principalMatch = _checkPrincipal(canonicalID, arn, s.Principal);
+        const actionMatch = _checkAction(requestType, s.Action);
+
+        if (principalMatch && actionMatch && s.Effect === 'Deny') {
+            explicitDeny = true;
```

Comment:
Yeah, I'm not a huge fan of mixing return types either, but I thought it would simplify the return of the calling function. I'll change it to keep the return consistent